RyuSocks.Test/Auth/UsernameAndPasswordTests.cs

```diff
@@ -0,0 +1,146 @@
+{
+    public class UsernameAndPasswordTests
+    {
+        [Theory]
+        [InlineData(new byte[] { 0xFF, 0xFF, 0xAA, 0x00, 0xCC })]
+        public void Authenticate_PackageWithWrongVersion(byte[] incomingPacket)
+        {
+            UsernameAndPassword usernameAndPassword = new()
+            {
+                Database = [],
+                IsClient = false
+            };
+            Assert.Throws<InvalidOperationException>(() => usernameAndPassword.Authenticate(incomingPacket, out _));
+        }
+
+        [Theory]
+        [UsernameAndPasswordRandomUsernameAndPasswords(1)]
+        public void Authenticate_RegisteredUserAuthentication(string username, string password)
+        {
+            UsernameAndPassword usernameAndPassword = new()
+            {
+                Database = new Dictionary<string, string> { { username, password } },
+                Username = username,
+                Password = password
+            };
+            usernameAndPassword.Authenticate(null, out ReadOnlySpan<byte> outgoingPacket);
+            usernameAndPassword.IsClient = false;
+
+            usernameAndPassword.Authenticate(outgoingPacket, out _);
+        }
+
+        [Theory]
+        [UsernameAndPasswordRandomUsernameAndPasswords(1)]
+        public void Authenticate_NotRegisteredUserAuthentication(string username, string password)
+        {
+            UsernameAndPassword usernameAndPassword = new()
+            {
+                Database = [],
+                Username = username,
+                Password = password
+            };
+            usernameAndPassword.Authenticate(null, out ReadOnlySpan<byte> outgoingPacket);
+            // Since a ByRef variable (of type ReadOnlySpan) is not accepted by a lambda expression, it is converted to a byte array
+            byte[] outgoingPacketByte = outgoingPacket.ToArray();
+            usernameAndPassword.IsClient = false;
+            Assert.Throws<AuthenticationException>(() => usernameAndPassword.Authenticate(outgoingPacketByte, out _));
+        }
+
+        [Theory]
+        [InlineData("RegisteredUser", "WrongPassword")]
+        public void Authenticate_RegisteredUserWrongPasswordAuthentication(string username, string password)
+        {
+            UsernameAndPassword usernameAndPassword = new()
+            {
+                Database = new Dictionary<string, string> { { username, "RightPassword" } },
+                Username = username,
+                Password = password
+            };
+            usernameAndPassword.Authenticate(null, out ReadOnlySpan<byte> outgoingPacket);
+            // Since a ByRef variable (of type ReadOnlySpan) is not accepted by a lambda expression, it is converted to a byte array
+            byte[] outgoingPacketByte = outgoingPacket.ToArray();
+            usernameAndPassword.IsClient = false;
+            Assert.Throws<AuthenticationException>(() => usernameAndPassword.Authenticate(outgoingPacketByte, out _));
+        }
```

Comment:
Instead of copying the method every time, you can adapt it so you can use with the new inputs as well.
Look at this test as an example: https://github.com/TSRBerry/RyuSOCKS/blob/7932e1a26d2482cb08e7aa782dbfa715a785bdbf/RyuSocks.Test/Packets/MethodSelectionRequestTests.cs#L92-L113